presto-main/src/main/java/com/facebook/presto/execution/ISqlTaskExecution.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution;
+
+import com.facebook.presto.TaskSource;
+import com.facebook.presto.execution.executor.TaskExecutor;
+import com.facebook.presto.operator.TaskContext;
+import com.facebook.presto.sql.planner.plan.PlanNodeId;
+
+import java.util.List;
+import java.util.Set;
+
+public interface ISqlTaskExecution
+{
+    void addSources(List<TaskSource> sources);
+
+    String toString();
+
+    TaskId getTaskId();
+
+    TaskContext getTaskContext();
+
+    Set<PlanNodeId> getNoMoreSplits();
+
+    static void addStateChangeListenerForTaskExecutor(TaskStateMachine taskStateMachine, TaskExecutor taskExecutor)
+    {
+        // A listener for updating failed tasks.
+        taskStateMachine.addStateChangeListener(state -> {
+            if (state.isDone()) {
+                taskExecutor.updateFailedTasks(state);
```

Comment:
This is a very specific detail to include in the interface. It should accept a generic listener.

Better yet, we can add the update operation into removeTask in TaskExecutor and avoid this new event listener

Reply:
I thought about doing that earlier, but that way we will miss out the tasks that fail before the listener for removeTask is registered.